packages/flutter_tools/lib/src/base/analyze_size.dart

```diff
@@ -0,0 +1,336 @@
+  }
+
+  SymbolNode _parent;
+  SymbolNode get parent => _parent;
+
+  final Map<String, SymbolNode> _children;
+
+  Iterable<SymbolNode> get children => _children.values;
+
+  SymbolNode childByName(String name) => _children[name];
+
+  SymbolNode addChild(SymbolNode child) {
+    assert(child.parent == null);
+    assert(!_children.containsKey(child.name),
+        'Cannot add duplicate child key ${child.name}');
+
+    child._parent = this;
+    _children[child.name] = child;
+    SymbolNode ancestor = this;
+    while (ancestor != null) {
+      ancestor._value += child.value;
+      ancestor = ancestor.parent;
+    }
+    return child;
+  }
+
+  List<SymbolNode> get ancestors {
+    final List<SymbolNode> nodes = <SymbolNode>[];
+    SymbolNode current = this;
+    while (current.parent != null) {
+      nodes.add(current.parent);
+      current = current.parent;
+    }
+    return nodes;
+  }
+
+  bool get isLeaf => _children.isEmpty;
+
+  Iterable<SymbolNode> get siblings {
+    final List<SymbolNode> result = <SymbolNode>[];
+    if (parent == null) {
+      return result;
+    }
+    for (final SymbolNode sibling in parent.children) {
+      if (sibling != this) {
+        result.add(sibling);
+      }
+    }
+    return result;
+  }
+}
+
+SymbolNode _parseSymbols(List<dynamic> symbols) {
+  final Iterable<Symbol> iter =
+      symbols.cast<Map<String, dynamic>>().map(Symbol.fromMap);
+  final SymbolNode root = SymbolNode('root');
+  SymbolNode currentParent = root;
+  for (final Symbol symbol in iter) {
+    final SymbolNode parentReset = currentParent;
+    for (final String pathPart in symbol.parts.take(symbol.parts.length - 1)) {
+      currentParent = currentParent.childByName(pathPart) ??
+          currentParent.addChild(SymbolNode(pathPart));
+    }
+    // TODO: this shouldn't be necessary, https://github.com/dart-lang/sdk/issues/41137
+    String leafName = symbol.parts.last;
+    int duplicates = 0;
+    while (currentParent.childByName(leafName) != null) {
+      duplicates += 1;
+      leafName = '${symbol.parts.last}_$duplicates';
+    }
+    currentParent.addChild(
+      SymbolNode(leafName, value: symbol.size),
+    );
+    currentParent = parentReset;
+  }
+  return root;
+}
+
+class Symbol {
+  const Symbol({
+    @required this.name,
+    @required this.size,
+    this.libraryUri,
+    this.className,
+  })  : assert(name != null),
+        assert(size != null);
+
+  static Symbol fromMap(Map<String, dynamic> json) {
+    return Symbol(
+      name: (json['n'] as String).replaceAll('[Optimized] ', ''),
+      size: json['s'] as int,
+      className: json['c'] as String,
+      libraryUri: json['l'] as String,
+    );
+  }
+
+  final String name;
+  final int size;
+  final String libraryUri;
+  final String className;
+
+  List<String> get parts {
+    return <String>[
+      if (libraryUri != null) ...libraryUri.split('/') else '@stubs',
+      if (className != null && className.isNotEmpty) className,
```

Comment:
nit: you can write
`className?.isNotEmpty ?? false` to avoid having to write `className` twice.